modules/evm/src/lib.rs

```diff
@@ -237,23 +271,121 @@ decl_module! {
+		{
+			CodeInfos::mutate_exists(&contract_info.code_hash, |maybe_code_info| {
+				if let Some(code_info) = maybe_code_info.as_mut() {
+					code_info.ref_count = code_info.ref_count.saturating_sub(1);
+					if code_info.ref_count == 0 {
+						Codes::remove(&contract_info.code_hash);
+						*maybe_code_info = None;
+					}
+				}
+			});
+		}
+
+		<Accounts<T>>::remove(address);
 		AccountStorages::remove_prefix(address);
 	}
 
 	/// Get the account basic in EVM format.
 	pub fn account_basic(address: &H160) -> Account {
 		let account_id = T::AddressMapping::to_account(address);
 
-		let nonce = Self::account_nonces(address);
+		let nonce = Self::accounts(address).map_or(Default::default(), |account_info| account_info.nonce);
 		let balance = T::Currency::free_balance(&account_id);
 
 		Account {
 			nonce: U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(nonce)),
 			balance: U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(balance)),
 		}
 	}
+
+	/// Get code hash at given address.
+	pub fn code_hash_at_address(address: &H160) -> H256 {
+		if let Some(AccountInfo {
+			contract_info: Some(contract_info),
+			..
+		}) = Self::accounts(address)
+		{
+			contract_info.code_hash
+		} else {
+			code_hash(&[])
+		}
+	}
+
+	/// Get code at given address.
+	pub fn code_at_address(address: &H160) -> Vec<u8> {
+		Self::codes(&Self::code_hash_at_address(address))
+	}
+
+	/// Handler on new contract initialization.
+	///
+	/// - Create new account for the contract.
+	///   - For contracts initialized in genesis block, `storage_count` param
+	///     needed to be provided.
+	///   - For contracts initialized via dispatch calls, storage count would be
+	///     read from initialized account storages.
+	/// - Update codes info.
+	/// - Save `code` if not saved yet.
+	pub fn on_contract_initialization(address: &H160, code: Vec<u8>, storage_count: Option<u32>) {
+		let code_hash = code_hash(&code.as_slice());
+		let storage_count = storage_count.unwrap_or_else(|| AccountStorages::iter_prefix(address).count() as u32);
+		let contract_info = ContractInfo {
+			storage_count,
+			code_hash,
+		};
+		Accounts::<T>::mutate(address, |maybe_account_info| {
+			if let Some(account_info) = maybe_account_info.as_mut() {
+				account_info.contract_info = Some(contract_info);
+			} else {
+				*maybe_account_info = Some(AccountInfo::<T::Index>::new(Default::default(), Some(contract_info)));
+			}
+		});
+
+		CodeInfos::mutate_exists(&code_hash, |maybe_code_info| {
+			if let Some(code_info) = maybe_code_info.as_mut() {
+				code_info.ref_count = code_info.ref_count.saturating_add(1);
+			} else {
+				let new = CodeInfo {
+					code_size: code.len() as u32,
+					ref_count: 1,
+				};
+				*maybe_code_info = Some(new);
+
+				Codes::insert(&code_hash, code);
+			}
+		});
+	}
+
+	/// Set account storage.
+	pub fn set_storage(address: H160, index: H256, value: H256) {
+		let mut is_remove = true;
+		if value == H256::default() {
```

Comment:
you need to read old value, and if already empty and set empty, it is a nop and no need to update storage count. same if value already exists and set to a value, also no need to update storage count.